cupy/fft/_callback.pyx

```diff
@@ -0,0 +1,450 @@
+                               '-I' + _cupy_include,
+                               '-fPIC', '-O2', '-std=c++11',
+                               '-c', tempdir + mod_name + '.cpp',
+                               '-o', obj_host],
+                               env=os.environ)
+            p.check_returncode()
+
+            # Dump and compile device code using nvcc
+            global _callback_dev_code
+            if _callback_dev_code is None:
+                with open(_source_dir + '/cupy_cufftXt.cu') as f:
+                    support = _callback_dev_code = f.read()
+            else:
+                support = _callback_dev_code
+            with open(tempdir + '/cupy_cufftXt.cu', 'w') as f:
+                support = string.Template(support).substitute(
+                    dev_load_callback_ker=cb_load,
+                    dev_store_callback_ker=cb_store)
+                f.write(support)
+            obj_dev = tempdir + mod_name + '_dev.o'
+            cmd = [_nvcc, '-ccbin', _cc[0],
+                   '-arch=sm_'+arch, '-dc',
+                   '-I' + _cupy_include,
+                   '-c', tempdir + '/cupy_cufftXt.cu',
+                   '-Xcompiler', '-fPIC', '-O2', '-std=c++11']
+            if self.cb_load:
+                cmd.append('-DHAS_LOAD_CALLBACK')
+            if self.cb_store:
+                cmd.append('-DHAS_STORE_CALLBACK')
+            p = subprocess.run(cmd + ['-o', obj_dev],
+                               env=os.environ,
+                               stdout=subprocess.PIPE,
+                               stderr=subprocess.PIPE)
+            try:
+                p.check_returncode()
+            except subprocess.CalledProcessError as e:
+                cex = CompileException(
+                    str(e) + '\nStderr: ' + e.stderr.decode(), support,
+                    tempdir + '/cupy_cufftXt.cu',
+                    cmd[1:], 'nvcc')
+                dump = _get_bool_env_variable(
+                    'CUPY_DUMP_CUDA_SOURCE_ON_ERROR', False)
+                if dump:
+                    cex.dump(sys.stderr)
+                raise cex
+
+            # Use nvcc to link and generate a shared library, and place it in
+            # the disk cache
+            # WARNING: CANNOT use host compiler to link!
+            p = subprocess.run([_nvcc, '-ccbin', _cc[0],
+                                '-shared', '-arch=sm_'+arch,
+                                obj_dev, obj_host,
+                                '-lcufft_static', '-lculibos', '-lpthread',
+                                '-o', path],
+                               env=os.environ)
+            p.check_returncode()
+
+            # Clean up build directory
+            tempdir_obj.cleanup()
+
+        # Load the Python module
+        spec = importlib.util.spec_from_file_location(mod_name, path)
+        self.mod = module = importlib.util.module_from_spec(spec)
+        sys.modules[mod_name] = module
+        spec.loader.exec_module(module)
```

Comment:
This is one of the most awful hacks I have ever seen, but I wouldn't be able to find a better approach myself 😂. Let's go ahead with this. 😜

Reply:
Yeah I know...I didn't like it either...But this is the only way to make it work.

Reply:
I wonder if we can refactor all these logic in some kind of cupy.utils.cython_module_loader class, maybe we will need something similar in the future?

Reply:
I am not sure if there's any other use case for this capability. In addition, the compilation steps are kinda specific to cuFFT's situation, so it's hard to make it general by nature. Maybe we will know better when the next case emerges?

Reply:
I guess there's a memory leak here? (Is it possible to remove completed callback from `sys.modules`?)

Reply:
I am not sure I follow the leak issue 😅 The idea is to use the same module for any kind of FFT transforms as long as they use the same callbacks, so the loaded modules are unloaded only at shutdown. For different sets of callbacks, the module is recompiled with a different name, so there won't be any import conflicts (i.e., there is no such a thing as "re-importing a callback module"). Does this address your concern @kmaehashi?

Reply:
BTW, this method is too long and hard to follow, can you split it in several private functions?

Reply:
Done in bf7efaa.